File: C#/210. Course Schedule II.cs
/*
 * Link: https://leetcode.com/problems/course-schedule-ii/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    // use topology sort
    public int[] FindOrder(int numCourses, int[][] prerequisites) {
        buildMap(numCourses, prerequisites);
        
        var st = TopologySort(numCourses);
        
        Console.WriteLine(st.Count);
        if (st.Count != numCourses) return new int[0];
        
        var res = new int[numCourses];
        for (var i = 0; i < numCourses; i++) {
            res[i] = st.Pop();
        }
        
        return res;
    }
    
    Dictionary<int, List<int>> map;
    public void buildMap(int numCourses, int[][] prerequisites) {
        map = new Dictionary<int, List<int>>();
        
        for(var i = 0; i < numCourses; i++) map[i] = new List<int>();
        
        foreach(var item in prerequisites) {
            map[item[1]].Add(item[0]);
        }
    }
    
    public Stack<int> TopologySort(int numCourses) {
        var st = new Stack<int>();
        var visited = new int[numCourses];
        Array.Fill(visited, 0);
        
        for(var i = 0; i < numCourses; i++) {
            if (visited[i] == 0) {
                if (!TopologySortUtil(i, ref visited, ref st)) return new Stack<int>();
            }
        }
        
        return st;
    }
    
    public bool TopologySortUtil(int current, ref int[] visited, ref Stack<int> st) {
        visited[current] = 1;
        
        foreach(var nextCourse in map[current]) {
            if (visited[nextCourse] == 1) return false;
            if (visited[nextCourse] == 0) {
                if(!TopologySortUtil(nextCourse, ref visited, ref st)) {
                    return false;
                };
            }
        }
        
        visited[current] = 2;
        st.Push(current);
        
        return true;
    }
}

File: Javascript/210. Course Schedule II.js
/**
 * Link: https://leetcode.com/problems/course-schedule-ii/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    var order = [];
    var map = createMap(numCourses, prerequisites);
    
    while (order.length < numCourses) {
        var check = false;
        for (var c in map) {
            var prerequisites = map[c];
            if (!order.includes(parseInt(c)) && hasLearnt(order, map, c)) {
                order.push(parseInt(c));
                check = true;
            }
        }
        
        // if go through the list of courses, cant find the next courses to learn, return false
        if (!check) {
            return [];
        }
    }
    
    return order;
}

// check if the prerequistics has learnt or not
function hasLearnt(order, map, c) {
    var prerequisites = map[c];
    
    if (prerequisites.length == 0 ||
       (order.length > 0 && prerequisites.every(p => order.includes(p)))) return true;
    
    return false;
}

// for every course, build the list of prerequisties
function createMap(numCourses, prerequisites) {
    var map = {};
    
    for (var i = 0; i < numCourses; i++) {
        map[i] = [];
    }
    
    for (var p of prerequisites) {
        map[p[0]].push(p[1]);
    }
    
    return map;
}